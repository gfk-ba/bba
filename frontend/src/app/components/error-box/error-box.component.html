<div class="row" *ngIf="error">
  <div class="col-12">
    <div class="error-box box">
      <div class="box-title">{{ title }}</div>
      <p>Status: {{ error.status }}</p>
      <p>Message: {{ error.message }}</p>
    </div>
  </div>
</div>
