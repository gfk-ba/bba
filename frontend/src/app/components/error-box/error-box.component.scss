@import '../../../styles/common/variables';

.error-box {
  border: 2px solid $red;
  color: $red;
  margin-top: $default-margin;
  margin-bottom: $default-margin;
}
