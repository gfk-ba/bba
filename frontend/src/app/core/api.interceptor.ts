import { Injectable } from '@angular/core';
import { Reset } from '@dcs/ngx-tools';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { catchError, flatMap, map, take } from 'rxjs/operators';
import { State } from '../reducers';
import { tokenSelector } from '../reducers/login/login.selectors';
import {
  HttpEvent,
  HttpHandler,
  HttpInterceptor,
  HttpRequest,
  HttpErrorResponse,
} from '@angular/common/http';

@Injectable()
export class ApiInterceptor implements HttpInterceptor {
  constructor(private store: Store<State>) {}

  intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    return this.store.pipe(
      take(1),
      map(state => tokenSelector(state)),
      flatMap(token => {
        request = request.clone({
          setHeaders: {
            Authorization: `Bearer ${token.accessToken}`,
          },
        });
        return next.handle(request).pipe(
          catchError((e: HttpErrorResponse) => {
            if (e.status === 401) {
              this.store.dispatch(new Reset());
            }
            throw e;
          })
        );
      })
    );
  }
}
