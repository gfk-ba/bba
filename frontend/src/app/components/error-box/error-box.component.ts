import { HttpErrorResponse } from '@angular/common/http';
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';

@Component({
  selector: 'dcs-error-box',
  templateUrl: './error-box.component.html',
  styleUrls: ['./error-box.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ErrorBoxComponent {
  @Input() public error: HttpErrorResponse;
  public title = 'Error';
}
