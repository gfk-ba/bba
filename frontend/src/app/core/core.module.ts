import { registerLocaleData } from '@angular/common';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import localeDe from '@angular/common/locales/de';
import { InjectionToken, LOCALE_ID, NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { ServiceWorkerModule } from '@angular/service-worker';
import { APP_ENVIRONMENT } from '@dcs/ngx-tools';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
import { ActionReducerMap, StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { ApiInterceptor } from './api.interceptor';
import { environment } from '../../environments/environment';
import { metaReducers, reducers, State } from '../reducers';
import { LoginGuard } from '../reducers/login/login.guard';

export const REDUCER_TOKEN = new InjectionToken<ActionReducerMap<State>>('Registered Reducers');

export function getReducers() {
  return reducers;
}

@NgModule({
  imports: [
    BrowserModule,
    HttpClientModule,
    StoreModule.forRoot(REDUCER_TOKEN, { metaReducers }),
    !environment.production ? StoreDevtoolsModule.instrument() : [],
    StoreRouterConnectingModule.forRoot({
      stateKey: 'router',
    }),
    ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production }),
  ],
  declarations: [],
  providers: [
    { provide: LOCALE_ID, useValue: 'en' },
    { provide: APP_ENVIRONMENT, useValue: environment },
    {
      provide: REDUCER_TOKEN,
      useFactory: getReducers,
    },
    {
      provide: HTTP_INTERCEPTORS,
      useClass: ApiInterceptor,
      multi: true,
    },
    LoginGuard,
  ],
})
export class CoreModule {
  constructor() {
    registerLocaleData(localeDe);
  }
}
