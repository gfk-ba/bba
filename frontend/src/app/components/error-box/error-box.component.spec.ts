import { Component, NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ErrorBoxComponent } from './error-box.component';
import { configureComponentTestSuite } from '../../../testing';

@Component({
  template: `<dcs-error-box [error]="error"></dcs-error-box>`,
})
class TestHostComponent {
  public error: { status: number; message: string };
}

describe('ErrorBoxComponent', () => {
  configureComponentTestSuite();

  beforeAll(async () => {
    await TestBed.configureTestingModule({
      declarations: [ErrorBoxComponent, TestHostComponent],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();
  });

  describe('testing the component instance', () => {
    let fixture: ComponentFixture<ErrorBoxComponent>;
    let comp: ErrorBoxComponent;

    beforeEach(() => {
      fixture = TestBed.createComponent(ErrorBoxComponent);
      comp = fixture.componentInstance;
    });

    it('can load instance', () => {
      expect(comp).toBeTruthy();
    });

    it('title defaults to: Error', () => {
      expect(comp.title).toEqual('Error');
    });
  });

  describe('testing the template output', () => {
    let fixture: ComponentFixture<TestHostComponent>;
    let comp: TestHostComponent;
    let errorEl: HTMLDivElement;

    beforeEach(() => {
      fixture = TestBed.createComponent(TestHostComponent);
      fixture.detectChanges();

      comp = fixture.componentInstance;
      errorEl = fixture.nativeElement.querySelector('dcs-error-box');
    });

    it('renders the empty error box component', () => {
      expect(errorEl).toMatchSnapshot();
    });

    it('renders the error if present', () => {
      comp.error = { status: 403, message: 'Keep out!' } as any;
      fixture.detectChanges();

      expect(errorEl).toMatchSnapshot();
    });
  });
});
