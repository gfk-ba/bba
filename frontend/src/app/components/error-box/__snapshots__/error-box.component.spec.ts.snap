// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoxComponent testing the template output renders the empty error box component 1`] = `
<dcs-error-box>
  
</dcs-error-box>
`;

exports[`ErrorBoxComponent testing the template output renders the error if present 1`] = `
<dcs-error-box
  ng-reflect-error="[object Object]"
>
  
  <div
    class="row"
  >
    <div
      class="col-12"
    >
      <div
        class="error-box box"
      >
        <div
          class="box-title"
        >
          Error
        </div>
        <p>
          Status: 403
        </p>
        <p>
          Message: Keep out!
        </p>
      </div>
    </div>
  </div>
</dcs-error-box>
`;
